quint/test/runtime/compile.test.ts

```diff
@@ -67,19 +67,25 @@ function assertVarExists(kind: ComputableKind, name: string, input: string) {
 
 // Scan the context for a callable. If found, evaluate it and return the value of the given var.
 // Assumes the input has a single callable
-function evalVarAfterCall(varName: string, input: string): Either<string, string> {
+function evalVarAfterCall(varName: string,
+    callee: string, input: string): Either<string, string> {
   // use a combination of Maybe and Either.
   // Recall that left(...) is used for errors,
   // whereas right(...) is used for non-errors in sweet monads.
   const callback = (ctx: CompilationContext): Either<string, string> => {
-    const key = [...ctx.values.keys()].find(k => k.startsWith('callable') && !builtinContext().has(k))
+    let key = undefined
+    if (ctx.main) {
+      const def = ctx.main.defs.find(def => def.kind === 'def' && def.name === callee)
+      if (def) {
+        key = kindName('callable', def.id)
+      }
+    }
     if (!key) {
-      return left('No callable found')
+      return left(`${callee} not found`)
     }
-
     const run = ctx.values.get(key)
     if (!run) {
-      return left(`${key} not found`)
+      return left(`${callee} not found via ${key}`)
     }
```

Comment:
The control flow here is pretty convoluted, iiuc:

Whether or not we can do anything in this function seems to depend on `ctx.main`: but that's not apparent unless the reader first reasons out the mutable update to `key` in combination with the early return.

Using early exits consistently  for the error conditions would clean this up a lot IMO:

```typescript
    if (!ctx.main) {
      return left('main module not found')
    }

    const def = ctx.main.defs.find(def => def.kind === 'def' && def.name === callee)
    if (!def) {
      return left(`${callee} deff not found`)
    }

    const key = kindName('callable', def.id)
    if (!key) {
      return left(`${callee} not found`)
    }

    const run = ctx.values.get(key)
    if (!run) {
      return left(`${callee} not found via ${key}`)
    }
```

Alternatively, one of the points of the Either data type is to sequence error prone computations that need to return some data on error. Since this is already returning an `Either`, we could make use of that here with a helper like:

```typescript
function definedOrErr<T>(x: T | undefined, err: string): Either<string, NonNullable<T>> {
  return x == null ? left(err) : right(x);
}
```

To pipeline the error handling and make the data dependencies very clear:

```typescript
function evalVarAfterCall(
  varName: string,
  callee: string,
  input: string
): Either<string, string> {
  // use a combination of Maybe and Either.
  // Recall that left(...) is used for errors,
  // whereas right(...) is used for non-errors in sweet monads.
  const callback = (ctx: CompilationContext): Either<string, string> => {
    return definedOrErr(ctx.main, "main module not defined")
      .chain((main) =>
        definedOrErr(
          main.defs.find((def) => def.kind === "def" && def.name === callee),
          `def for ${callee} not found`
        )
      )
      .chain((def) =>
        definedOrErr(kindName("callable", def.id), `${callee} not found`)
      )
      .chain((key) =>
        definedOrErr(ctx.values.get(key), `${callee} not found via ${key}`)
      )
      .chain((run) =>
        run
          .eval()
          .map((res) => {
            if ((res as RuntimeValue).toBool() === true) {
              // extract the value of the state variable
              const nextVal = (
                ctx.values.get(kindName("nextvar", varName)) ?? fail
              ).eval();
              // using if-else, as map-or-unwrap confuses the compiler a lot
              if (nextVal.isNone()) {
                return left(`Value of the variable ${varName} is undefined`);
              } else {
                return right(
                  expressionToString(nextVal.value.toQuintEx(idGen))
                );
              }
            } else {
              const s = expressionToString(res.toQuintEx(idGen));
              const m = `Callable ${callee} was expected to evaluate to true, found: ${s}`;
              return left<string, string>(m);
            }
          })
          .or(just(left(`Value of ${callee} is undefined`)))
          .unwrap()
      );
  };

  return evalInContext(input, callback);
}
```

Reply:
Yeah, the either chaining doesn’t add a lot of value here afaict :)